### DISCUSSION TITLE
[Bug]: code-server hang

### DISCUSSION THREAD

**Discussion by ericzhucode** [coder/code-server#5822] [Q&A]
### Is there an existing issue for this?

- [X] I have searched the existing issues

I know this issue may not be caused by code-server but nodejs since when process hangs, it seems no any js code is running and all nodejs environment stucks. Just post this here to see if it's possible to get professional suggestions from your team.
I will also post this issue at nodejs community
Thank you for your team's effort!

### OS/Web Information

- Web Browser: Chrome/Edge(This issue is only related to backend process on Linux system)
- Local OS: Win10/Win11
- Remote OS: Linux CentOS
- Remote Architecture: 
- `code-server --version`:  4.1.0 / 4.2.0 / 4.4.0 (Some customers are upgrading to 4.5.3 now and we will monitor if this version also has the issue)

### Steps to Reproduce

1. start code-server
2. use code-server as usual
3. after about one week to one month, the code-server hangs
4. we can't reproduce this issue manually as this always happens suddenly and unexpectedly

Troubleshooting so far
1. We can't open code-server in browser as chrome shows no response at all, but port access is still avaiable when we use telnet
2. We can find code-server process  still exists and it is still listening on the port (with `telnet 127.0.0.1 8080`)
![process](https://user-images.githubusercontent.com/85340915/203211478-974a5aef-daf5-425c-a261-1b6a87aaf896.png)
3. We even can't curl on the local linux machine and there's no response at all
![noresponse](https://user-images.githubusercontent.com/85340915/203240266-f5b5ff8b-9bd4-44fa-b3d1-f901c1744a9e.png)

(if it's working then we should see the direction infomation like)
![redirect](https://user-images.githubusercontent.com/85340915/203240258-0aa36fc8-901c-4660-9dbc-4bda5ca154c9.png)

4. We generated core file of the hang process (I mean the process executing `entry.js`) and use gdb to debug it. When we use `bt` to look at the frames, we can get the following content
![core file](https://user-images.githubusercontent.com/85340915/203240530-878e177d-9ea1-4431-9a63-bdd8f7197972.png)
We get this core files from over 20+ of our users and the contents are all the same, showing nodejs is stucking at
write() from /lib64/libpthread.so.0

### Expected

nodejs server in linux backend system should response when we use curl or open IDE in browser

### Actual

nodejs process in linux backend system hangs and no response at all

### Logs

from ~/.local/share/code-server/code-server-stderr.log
We found the following logs and it seems that there's network traffic issue in socket connection but I don't know how we use these information to troubleshoot. I will be appreciated it if you team can suggest something.
![image](https://user-images.githubusercontent.com/85340915/203241581-a7e64062-54e0-4efe-a1d6-4b3ba15f64fc.png)

### Does this issue happen in VS Code or GitHub Codespaces?

- [X] I cannot reproduce this in VS Code.
- [X] I cannot reproduce this in GitHub Codespaces.

### Are you accessing code-server over HTTPS?

- [X] I am using HTTPS.

**Comment by jsjoeio** [comment]
This is a tough one 🤔 Nothing obvious is coming to me after reading your logs. @code-asher any ideas? I'm not sure if long-running processes have issues or something.

What happens if you restart code-server? Does it work then?

**Reply by ericzhucode** [reply]
Hi! Thank you for your reply.
Yes, this is really a tough one and hard to troubleshoot where exactly the reason is, but after posting this issue on nodejs community, we recived some hints from them and here's the link.
https://github.com/nodejs/help/issues/4039
It seems the main process which executes entry.js stucks because the stdout pipe is full. According to this, we have checked the user's log file which code-server outputs to
~/.local/share/code-server/coder-logs/
and we found there's no any log at all until we restarted server. Besides, the log expected to be outputed (which hasn't been outputed exactly) did displayed in `strace -yy -p <pid>` suggested by nodejs team.

I think this all of these evidence can prove that the pipe is full and nodejs stucks at this point.

However, we have suggested our users to update to version 4.5.3 and from last week, the number of this kind of issue extreamly decreased.
I'm not sure if it's because @code-asher has modified the logic of log output, removing `stdout.pipe` and use `stdout.on`
https://github.com/coder/code-server/pull/5346
![image](https://user-images.githubusercontent.com/85340915/205540326-76743319-9f44-4177-b4a9-09dc90a54b64.png)

Currently I think this issue has been resolved since v4.5.3, though I'm not quite sure.

**Reply by jsjoeio** [reply]
Very very interesting! @code-asher returns from PTO later this week so maybe he'll have some insight into this.

**Reply by code-asher** [reply]
Fascinating!  Yeah there was a problem with stdout not being consumed although I only saw this after a hotswap (which code-server performs when it receives a SIGUSR1 or SIGUSR2) so I am not sure how the issue would trigger without a hotswap but if it does that would explain it.

**Reply by ericzhucode** [reply]
> Fascinating! Yeah there was a problem with stdout not being consumed although I only saw this after a hotswap (which code-server performs when it receives a SIGUSR1 or SIGUSR2) so I am not sure how the issue would trigger without a hotswap but if it does that would explain it.

Can I ask more about `hotswap` you mentioned above? I know signal SIGUSR1 and SIGUSR2 since we sometimes use that in order to do an attach debug, but what does hotswap mean? Is it a feature controlled by code-server code or by express frameworks or nodejs?

**Reply by code-asher** [reply]
In this case I mean code-server restarts itself.  code-server's top-level process is a thin wrapper that starts the main code-server process and when the top-level code-server process receives `SIGUSR1` or `SIGUSR2` it restarts the main code-server process.  I think in the `pipe` implementation when the main process streams closed they closed the top-level ones too.

**Reply by code-asher** [reply]
We should probably remove `SIGUSR1` since that overlaps with debugging Node.

**Reply by ericzhucode** [reply]
Much appreciated for your insights!

**Comment by kuelabs** [comment]
I've had the same problem

code-server: v4.102.1
Code: 1.102.1
Commit: 47e9d43922b5764e3ead3bbf31d9ad23676604d6
Date: 2025-07-17T18:59:27.470Z (2 mos ago)
Browser: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36 Edg/140.0.0.0
